File: backend/src/application/services/support_case_service.py
"""Support Case Service for coordinating support case operations."""

from typing import List, Dict, Any
from uuid import UUID
from datetime import datetime

from ..use_cases.create_support_case import CreateSupportCase
from ..use_cases.get_support_case import GetSupportCase
from ..use_cases.update_support_case_status import UpdateSupportCaseStatus
from ...domain.repositories.support_case_repository import SupportCaseRepository


class SupportCaseService:
    """Service layer for support case operations."""

    def __init__(self, repository: SupportCaseRepository):
        self.repository = repository
        self.create_case_use_case = CreateSupportCase(repository)
        self.get_case_use_case = GetSupportCase(repository)
        self.update_status_use_case = UpdateSupportCaseStatus(repository)

    def create_support_case(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new support case."""
        return self.create_case_use_case.execute(request)

    def get_support_case(self, case_id: UUID) -> Dict[str, Any]:
        """Get a support case by ID."""
        return self.get_case_use_case.execute(case_id)

    def update_support_case_status(self, case_id: UUID, status: str) -> Dict[str, Any]:
        """Update support case status."""
        return self.update_status_use_case.execute(case_id, status)

    def get_customer_support_cases(self, customer_id: UUID) -> List[Dict[str, Any]]:
        """Get all support cases for a customer."""
        cases = self.repository.find_by_customer_id(customer_id)

        return [
            {
                "case_id": str(case.case_id),
                "customer_id": str(case.customer_id),
                "order_id": str(case.order_id),
                "title": case.title,
                "description": case.description,
                "status": case.status,
                "created_at": case.created_at.isoformat(),
                "updated_at": case.updated_at.isoformat(),
            }
            for case in cases
        ]

File: frontend/src/services/supportCaseService.ts
import axios from 'axios'

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000'

// Create axios instance with default config
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
})

// Add auth token interceptor
apiClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('auth_token')
    if (token) {
      config.headers.Authorization = `Bearer ${token}`
    }
    return config
  },
  (error) => {
    return Promise.reject(error)
  }
)

export type SupportCase = {
  case_id: string
  customer_id: string
  order_id: string
  title: string
  description: string
  status: 'open' | 'in_progress' | 'resolved' | 'closed'
  created_at: string
  updated_at: string
}

export type CreateSupportCaseRequest = {
  title: string
  description: string
  orderId: string
}

export type UpdateSupportCaseStatusRequest = {
  status: 'open' | 'in_progress' | 'resolved' | 'closed'
}

class SupportCaseService {
  async createSupportCase(request: CreateSupportCaseRequest): Promise<SupportCase> {
    try {
      const response = await apiClient.post('/api/support-cases', request)
      return response.data
    } catch (error) {
      throw new Error(`Failed to create support case: ${error}`)
    }
  }

  async getSupportCase(caseId: string): Promise<SupportCase> {
    try {
      const response = await apiClient.get(`/api/support-cases/${caseId}`)
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch support case: ${error}`)
    }
  }

  async updateSupportCaseStatus(
    caseId: string,
    status: UpdateSupportCaseStatusRequest
  ): Promise<SupportCase> {
    try {
      const response = await apiClient.patch(
        `/api/support-cases/${caseId}/status`,
        status
      )
      return response.data
    } catch (error) {
      throw new Error(`Failed to update support case status: ${error}`)
    }
  }

  async getMySupportCases(): Promise<SupportCase[]> {
    try {
      const response = await apiClient.get('/api/support-cases')
      return response.data
    } catch (error) {
      throw new Error(`Failed to fetch support cases: ${error}`)
    }
  }
}

export const supportCaseService = new SupportCaseService()